src/util/audio/AudioPlayer.cpp

```diff
@@ -0,0 +1,93 @@
+#include "AudioPlayer.h"
+
+AudioPlayer::AudioPlayer(Settings *settings) : settings(settings)
+{
+    XOJ_INIT_TYPE(AudioPlayer);
+
+    this->audioQueue = new AudioQueue();
+    this->portAudioConsumer = new PortAudioConsumer(settings, this->audioQueue);
+    this->soxProducer = new SoxProducer(this->audioQueue);
+}
+
+AudioPlayer::~AudioPlayer()
+{
+    XOJ_CHECK_TYPE(AudioPlayer);
+
+    delete this->portAudioConsumer;
+    this->portAudioConsumer = nullptr;
+
+    delete this->soxProducer;
+    this->soxProducer = nullptr;
+
+    delete this->audioQueue;
+    this->audioQueue = nullptr;
+
+    XOJ_RELEASE_TYPE(AudioPlayer);
+}
+
+void AudioPlayer::start(string filename, unsigned int timestamp)
+{
+    XOJ_CHECK_TYPE(AudioPlayer);
+
+    // Start the producer for reading the data
+    this->soxProducer->start(std::move(filename), this->portAudioConsumer->getSelectedOutputDevice(), timestamp);
+    sox_signalinfo_t *signal = this->soxProducer->getSignalInformation();
+
+    // Start playing
+    if (signal->rate > 1)
+    {
+        this->portAudioConsumer->startPlaying(signal->rate, signal->channels);
+    }
+    else
+    {
+        this->portAudioConsumer->startPlaying(44100.0, signal->channels);
+    }
+
+    // Clean up after audio is played
+    stopThread = std::thread([&] {
+        while(this->portAudioConsumer->isPlaying())
+        {
+            Pa_Sleep(100);
+        }
+        this->stop();
+    });
+    //TODO we need something nicer here then detaching the thread
```

Comment:
Is there no listener or something like this?

Reply:
You have an example in your code: Lambdas have a closure, so it should possible to use this to additional add all needed data.
(before lambdas: mem_fun can do this also)

I don't have a working example, but I'm really sure it will work ;-)